### PR TITLE
user listing with sorting and parameter for search

### PR DESCRIPTION
**Pull request by ps-eqsdigital**
sort must be 0 or 1
if 0 - sort by ascending name
if 1 - sort by descending name
if sort isnt 0 or 1 - return error 400

search parameter to search for name or emaill
didnt add "category" parameter because i didnt find much sense in that (talked with ricardo)

### CODE REVIEW

File: api/Business/BusinessObjects/UserBusinessObject.cs [hunk 1]
```diff
@@ -48,6 +50,19 @@ public async Task<OperationResult> Update(Guid uuid, User record)
...
 
+        public async Task<OperationResult<List<User>>> ListFilteredUsers(string search, int sort)
+        {
+            return await ExecuteOperation(async () => {
+
+                if (sort !=0 && sort != 1)
+                {
+                    throw new Exception();
+                }
+                var result =await _userDataAccessObject.FilterUsers(search,sort);
```
Comment (kripton69) [inline api/Business/BusinessObjects/UserBusinessObject.cs]: don't use "var" use the real type: for this case I think it's List<User>

File: api/DataAccess/DataAccessObjects/UserDataAccessObject.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        public UserDataAccessObject(UnoOnBoardingContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<List<User>> FilterUsers(string search,int sort)
+        {
+            if (sort == 0)
+            {
+                var result = await _context.Set<User>()
```
Comment (kripton69) [inline api/DataAccess/DataAccessObjects/UserDataAccessObject.cs]: types along file

File: api/DataAccess/DataAccessObjects/UserDataAccessObject.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                            (x.Email!.ToLower().Contains(search.ToLower())))
+                .OrderBy(x => x.Name).ToListAsync();
+
+                return result.Where(x => !x.IsDeleted).ToList(); ;
+            }
+
+            else
+            {
+                var result = await _context.Set<User>()
+                .Where(x => (x.Name!.ToLower().Contains(search.ToLower())) ||
```
Comment (kripton69) [inline api/DataAccess/DataAccessObjects/UserDataAccessObject.cs]: pay attention to words with accents: "Antônio" does not contain "Antonio"

File: api/api/Controllers/UserController.cs [hunk 4]
```diff
@@ -16,7 +16,7 @@ public UserController(IGenericBusinessObject genericBusinessObject, IUserBusines
         }
 
         [HttpGet("get")]
-        public async Task<ActionResult<List<User>>> ListUsers()
+        public async Task<ActionResult<List<User>>> GetUsers()
         {
             var result = await _genericBusinessObject.ListAsync<User>();
```
Comment (kripton69) [inline api/api/Controllers/UserController.cs:22]: types along file

**Review by js-eqsdigital — APPROVED** [review]

File: api/Business/BusinessObjects/UserBusinessObject.cs (hunk 1 above)
Comment (ps-eqsdigital) [reply api/Business/BusinessObjects/UserBusinessObject.cs]: donw

File: api/DataAccess/DataAccessObjects/UserDataAccessObject.cs (hunk 2 above)
Comment (ps-eqsdigital) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs]: done

File: api/DataAccess/DataAccessObjects/UserDataAccessObject.cs (hunk 3 above)
Comment (ps-eqsdigital) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs]: added extension method

File: api/api/Controllers/UserController.cs (hunk 4 above)
Comment (ps-eqsdigital) [reply api/api/Controllers/UserController.cs:22]: done

File: api/DataAccess/DataAccessObjects/UserDataAccessObject.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            _context = context;
+        }
+
+        public async Task<List<User>> FilterUsers(string search,int sort)
+        {
+            search = search.RemoveDiacritics();
+            if (sort == 0)
+            {
+                List<User> result = await _context.Set<User>()
+                .Where(x => (x.Name!.ToLower().Contains(search.ToLower())) ||
```
Comment (kripton69) [inline api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: Tens que remover no Name tbm, porque na bd pode estar o nome Antônio

Comment (ps-eqsdigital) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: nao consigo aplicar estes metodos a queries à BD, penso que a unica soluçao seja mesmo nas definiçoes da tabela da BD

Comment (kripton69) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: Mas não podes fazer isto:  x.Name!.ToLower().RemoveDiacritics().Contains(search.ToLower())????

Comment (ps-eqsdigital) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: não, pq nao dá para converter isso numa query sql
https://stackoverflow.com/questions/7418624/linq-where-ignore-accentuation-and-case

Comment (kripton69) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: @js-eqsdigital deixa aqui o teu conhecimento.

Comment (js-eqsdigital) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: @ps-eqsdigital  tens que definir a coluna string deste tipo Latin1_General_CI_AI para que seja Accent Insensitive.

**Review by js-eqsdigital — DISMISSED** [review]

File: api/DataAccess/DataAccessObjects/UserDataAccessObject.cs (hunk 5 above)
Comment (ps-eqsdigital) [reply api/DataAccess/DataAccessObjects/UserDataAccessObject.cs:30]: done

**Review by kripton69 — APPROVED** [review]

**Review by js-eqsdigital — APPROVED** [review]

**Pull request merged**